### ISSUE TITLE
[Linux/arm64] Infinite loop during IL_Rethrow

### ISSUE THREAD

**Issue by echesakovMSFT** [dotnet/coreclr#20314]
The following program (Release)
```c#
using System;

namespace CatchThrow
{
    class Program
    {
        static void Main(string[] args)
        {
            try
	    {
	        throw new Exception();
	    }
	    catch
	    {
	        throw;
	    }
        }
    }
}
```
will stuck in a loop during `IL_Rethrow` https://github.com/dotnet/coreclr/blob/913428d5915a9729c9405f57a75e7f912f9d29a5/src/vm/exceptionhandling.cpp#L4659-L4704

```
(lldb) bt
* thread #1, name = 'corerun', stop reason = signal SIGSTOP
  * frame #0: 0x0000ffffb7cb1b44 libc.so.6`syscall at syscall.S:39
    frame #1: 0x0000ffffb77aad7c libcoreclr.so`::GetCurrentThreadId() at thread.cpp:395
    frame #2: 0x0000ffffb71bb440 libcoreclr.so`ThreadStressLog::LogMsg(this=<unavailable>, facility=<unavailable>, cArgs=<unavailable>, format=<unavailable>, Args=<unavailable>) at stresslog.cpp:581
    frame #3: 0x0000ffffb71bb5fc libcoreclr.so`StressLog::LogMsg(level=<unavailable>, facility=16384, cArgs=2, format=0x0000ffffb7866882) at stresslog.cpp:665
    frame #4: 0x0000ffffb74302bc libcoreclr.so`UnwindManagedExceptionPass1(ex=<unavailable>, frameContext=<unavailable>) at exceptionhandling.cpp:4672
    frame #5: 0x0000ffffb7430640 libcoreclr.so`DispatchManagedException(ex=<unavailable>, isHardwareException=<unavailable>) at exceptionhandling.cpp:4752
    frame #6: 0x0000ffffb736c59c libcoreclr.so`IL_Rethrow() at jithelpers.cpp:4894
    frame #7: 0x0000ffff3d763980
    frame #8: 0x0000ffffb743941c libcoreclr.so`CallDescrWorkerInternal + 132
    frame #9: 0x0000ffffb72f05c8 libcoreclr.so`CallDescrWorkerWithHandler(pCallDescrData=0x0000ffffffffe5a8, fCriticalCall=<unavailable>) at callhelpers.cpp:78
    frame #10: 0x0000ffffb72f13b4 libcoreclr.so`MethodDescCallSite::CallTargetWorker(this=<unavailable>, pArguments=0x0000ffffffffe6f8, pReturnValue=0x0000000000000000, cbReturnValue=0) at callhelpers.cpp:629
    frame #11: 0x0000ffffb7461db8 libcoreclr.so`RunMain(MethodDesc*, short, int*, REF<PtrArray>*) [inlined] MethodDescCallSite::Call(this=<unavailable>, pArguments=<unavailable>) at callhelpers.h:472
    frame #12: 0x0000ffffb7461da4 libcoreclr.so`RunMain(MethodDesc*, short, int*, REF<PtrArray>*) at assembly.cpp:1660
    frame #13: 0x0000ffffb7461be8 libcoreclr.so`RunMain(MethodDesc*, short, int*, REF<PtrArray>*) at assembly.cpp:1675
    frame #14: 0x0000ffffb7461b28 libcoreclr.so`RunMain(pFD=<unavailable>, numSkipArgs=<unavailable>, piRetVal=<unavailable>, stringArgs=<unavailable>) at assembly.cpp:1675
    frame #15: 0x0000ffffb7462268 libcoreclr.so`Assembly::ExecuteMainMethod(this=<unavailable>, stringArgs=0x0000ffffffffec38, waitForOtherThreads=YES) at assembly.cpp:1785
    frame #16: 0x0000ffffb71d69c0 libcoreclr.so`CorHost2::ExecuteAssembly(this=<unavailable>, dwAppDomainId=<unavailable>, pwzAssemblyPath=<unavailable>, argc=<unavailable>, argv=<unavailable>, pReturnValue=<unavailable>) at corhost.cpp:484
    frame #17: 0x0000ffffb71a0cd8 libcoreclr.so`::coreclr_execute_assembly(hostHandle=<unavailable>, domainId=<unavailable>, argc=<unavailable>, argv=<unavailable>, managedAssemblyPath=<unavailable>, exitCode=<unavailable>) at unixinterface.cpp:407
    frame #18: 0x0000000000402f24 corerun`ExecuteManagedAssembly(currentExeAbsolutePath=<unavailable>, clrFilesAbsolutePath=<unavailable>, managedAssemblyAbsolutePath=<unavailable>, managedAssemblyArgc=<unavailable>, managedAssemblyArgv=<unavailable>) at coreruncommon.cpp:429
    frame #19: 0x0000000000402218 corerun`corerun(argc=<unavailable>, argv=<unavailable>) at corerun.cpp:149
    frame #20: 0x0000ffffb7c0e8a0 libc.so.6`__libc_start_main(main=0x0000000000000000, argc=0, argv=0x0000000000000000, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=<unavailable>) at libc-start.c:291
    frame #21: 0x0000000000401e58 corerun`_start + 40
```

**Environment:**
```
COMPlus_LogLevel=7
COMPlus_StressLog=1
COMPlus_LogFacility=0x00004000
COMPlus_JitDiffableDasm=0
COMPlus_TieredCompilation=0
COMPlus_TotalStressLogSize=40000000
COMPlus_JitEHDump=CatchThrow.Program:Main
COMPlus_JitDisasm=CatchThrow.Program:Main
COMPlus_StressLogSize=2000000
COMPlus_ZapDisable=1
COMPlus_ReadyToRun=0
```

**Comment by echesakovMSFT** [comment]
StressLog.txt
```
STRESS LOG:
    facilitiesToLog  = 0x80004000
    levelToLog       = 7
    MaxLogSizePerThread = 0x2000000 (33554432)
    MaxTotalLogSize = 0x40000000 (1073741824)
    CurrentTotalLogChunk = 2049
    ThreadsWithLogs  = 3
    Clock frequency  = 1.000 GHz
    Start time             Last message time      Total elapsed time 4.839 sec

THREAD  TIMESTAMP     FACILITY                              MESSAGE
  ID  (sec from start)
--------------------------------------------------------------------------------------
 e7d   4.839329844 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839328444 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839326994 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839325544 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839324044 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839322644 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839321194 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839319744 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839318245 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839316845 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839315395 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839313945 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839312445 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839311045 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839309595 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839308145 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839306696 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839305246 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839303796 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839302346 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839300846 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839299396 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839297996 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839296546 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060

 e7d   4.839295047 : `EH`                 Processing exception at native frame: IP = 0xffffb7c0e8a0, SP = 0xfffffffff060
```

**Comment by janvorli** [comment]
@echesakovMSFT that means that the libunwind cannot unwind the native frame #20` `libc.so.6`__libc_start_main` for some reason. It is either a C++ compiler bug or the libunwind bug. It should be relatively easy to check what's libunwind doing since its source is part of coreclr repo. I can look into that.

**Comment by janvorli** [comment]
So the issue is that the `__libc_start_main` is missing unwind info and it has non-trivial frame that cannot be unwound without that. It starts with 
```asm
```
I've found that the recent glibc versions actually have patch to fix that:
https://sourceware.org/git/?p=glibc.git;a=commit;h=ca3a382ea3aa9c9ccb57060afae47c7f702727ac
This change is in glibc 2.27 and newer.

The Ubuntu16.04 where you were running the test has glibc 2.23.

We already had a check to process similar case, but it gets triggered only if the `unw_step` returns 0 and the IP before and after the `unw_step` is the same. But in the arm64 case, the unw_step returns 1 (which still indicates success). So we should fix it by checking the return code to >=0 instead. I've verified it fixes the issue.
